source/opt/merge_return_pass.cpp

```diff
@@ -239,8 +239,21 @@ void MergeReturnPass::CreatePhiNodesForInst(BasicBlock* merge_block,
   if (inst.result_id() != 0) {
     std::vector<Instruction*> users_to_update;
     context()->get_def_use_mgr()->ForEachUser(
-        &inst, [&users_to_update, &dom_tree, inst_bb, this](Instruction* user) {
-          BasicBlock* user_bb = context()->get_instr_block(user);
+        &inst, [&users_to_update, &dom_tree, &inst, inst_bb, this](Instruction* user) {
+          BasicBlock* user_bb = nullptr;
+          if (user->opcode() != SpvOpPhi) {
+            user_bb = context()->get_instr_block(user);
+          } else {
+            // For OpPhi, the use should be considered to be in the predecessor.
+            for(uint32_t i = 0; i < user->NumInOperands(); i+=2) {
+              if (user->GetSingleWordInOperand(i) == inst.result_id()) {
+                uint32_t user_bb_id = user->GetSingleWordInOperand(i+1);
+                user_bb = context()->get_instr_block(user_bb_id);
+                break;
```

Comment:
What happens if the same value is merged across more than one predecessor? OpSwitch could generate an OpPhi with 3+ incoming edges with 2+ that use the same value.

Reply:
I thought of this.  We will replace all of the uses of that id.  The case where this can go wrong is if there is a case where we want to replace it in one but not the other.  I'll try harder to try to come up with a case like that.

Reply:
This part of the code should be fine.  I do have an informal proof of correctness.  We can discuss it in person.  However, I did find a bug in another part of the code.  I'll try to get that fixed up tomorrow and make it part of this PR.  It is similar.  We are not rewritting an OpPhi correctly.

Reply:
I've added a new test case and fixes for another case where OpPhis are getting messed up.  We can discuss if there are other cases that can go wrong.